### ISSUE TITLE
CircuitBreaker AutoTransitionner prevents JVM shutdown

### ISSUE THREAD

**Issue by 0xf10413** [resilience4j/resilience4j#351]
Hello,

I found what looks like a bug when using the AutoTransitionner. Take this code:
```
public static void main(String [] args) {
        final CircuitBreakerConfig circuitBreakerConfig = CircuitBreakerConfig.custom()
                .ringBufferSizeInClosedState(2)
                .failureRateThreshold(50.f)
                .waitDurationInOpenState(Duration.ofMillis(10000))
                .enableAutomaticTransitionFromOpenToHalfOpen()
                .build();
        
        final CircuitBreaker cb = CircuitBreaker.of("cb1", circuitBreakerConfig);
        Callable<String> wrappedBoom = CircuitBreaker.
                decorateCallable(cb, () -> {throw new RuntimeException("boom");});
        
        // Fail twice to open cb
        try {
            wrappedBoom.call();
        } catch (Exception e) {}
        try {
            wrappedBoom.call();
        } catch (Exception e) {}
        
        // Welp you're stuck now
        System.out.println("JVM will never close");
    }
```

If you follow that program in the debugger, you can see that a running thread prevents the JVM shutdown. My guess is that this is a consequence of the AutoTransitionner ExecutorService, which is started but never shutdown. We'd need to call shutdown+awaitTermination or (probably better) shutdownNow to close it.

As it is a private instance of an internal class, i don't think there's a way to solve this without modifying the lib. Maybe adding a new call to shutdown this executor ?

Apologies in advance if this is a duplicate. I skimmed through your opened issues but wasn't able to find a similar one.

**Comment by RobWin** [comment]
Hi
thx for reporting. Would you like to create a bugfix?

**Comment by 0xf10413** [comment]
Hi !
Sorry, i'm still not fully familiar with Java to know how to perform that. I thought of maybe adding a static shutdown function somewhere in the interface (ugly), or to add one AutoTransitionner per circuitbreaker and make the ExecutorService disappear with the circuitbreaker.
Unfortunately, just losing the reference to the executor service is not enough to close it...
I googled the AutoClosable interface, it would work with a try-with-resources maybe...
Or maybe change the pool type to allow no thread at all...
I'll try to take a peek this weekend.

**Comment by mumukiller** [comment]
I'm not sure that AutoClosable can help, as it should be triggered by some event (e.g. in case of spring boot autoconfiguration it can be method which is marked as @PreDestroy)

For manual CircuitBreaker management some method should be executed in order to prepare shutdown procedure.

So i suppose there should be separate method shutdown() in CircuitBreaker interface

Still there is a question regarding AutoTransitioner Executor. Is it ok to have Executor per CircuitBreaker or probably this should be configurable? @RobWin any ideas regarding this?

**Comment by storozhukBM** [comment]
Fixed. Will be released as part of 0.14.0
